feat(runner): Container detection and handling, video bit-depth flags and script updates

DatCaptainHorse opened the pull request:
## Description

Works in apptainer now.. podman is still the goat since apptainer needs docker treatment and even more..

- Added container detection so podman can be used to it's fullest, the non-sane ones are handled separately..
- Added video bit-depth option, cuz AV1 and 10-bit encoding go well together.
- Some other package updates to nestri-server.
- General tidying up of scripts to make multi-container-engine handling less of a pain.
- Updated old wireplumber lua script to new json format.

Further changes:

- Removed unused debug arg from nestri-server.
- Moved configs to config file folder rather than keeping them in containerfile.
- Improved audio configs, moved some into wireplumber to keep things tidy.
- Bit better arg handling in nestri-server.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Optional 10‑bit video support and auto‑launch of an app after display setup.

* **Changes**
  * Standardized runtime/user env to NESTRI_* with updated home/cache paths and explicit LANG; password generation now logged.
  * Improved container/GPU detection and startup logging; reduced blanket root usage during startup; SSH setup surfaced.
  * WirePlumber/PipeWire moved to JSON configs; low‑latency clock and loopback audio policies added; audio capture defaults to PipeWire.
  
* **Chores**
  * GStreamer/libp2p dependency upgrades and Rust toolchain pinned; NVIDIA driver capability exposed.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

DatCaptainHorse committed:
feat(runner): Improved container handling and updates

- Added container detection so podman can be used to it's fullest, the non-sane ones are handled separately..
- Added video bit-depth option, cuz AV1 and 10-bit encoding go well together.
- Some other package updates to nestri-server.
- General tidying up of scripts to make multi-container-engine handling less of a pain.

wanjohiryan commented on packages/scripts/entrypoint.sh:
```diff
@@ -307,20 +315,26 @@ main() {
-    # Setup namespaceless env
-    log "Applying namespace-less configuration"
-    setup_namespaceless
+    # Setup namespaceless env for non-sane container engines
```
😆 "non-sane container engines"

DatCaptainHorse replied:
I might have some beef with Docker (and now apptainer, since it requires same workarounds..) :sweat_smile: - we could have things better if everyone agreed on something, but no, cue the "standards" xkcd: https://xkcd.com/927/

DatCaptainHorse commented on containerfiles/runner.Containerfile:
```diff
@@ -121,58 +121,49 @@ RUN --mount=type=cache,target=${CARGO_HOME}/registry \
+    NESTRI_LANG=en_US.UTF-8 \
+    NESTRI_XDG_RUNTIME_DIR=/run/user/1000 \
+    NESTRI_HOME=/home/nestri \
     NVIDIA_DRIVER_CAPABILITIES=all
```
Fair enough, since it's not used anyhow from user-side, should be resolved in next commit.

DatCaptainHorse replied:
FYI, ARG and ENV do NOT evaluate commands, I'll be sending a fix-up commit which uses the RUN section below to generate the nestri user pass. Make sure you put it in your learnings, ARG and ENV do NOT evaluate commands.

DatCaptainHorse commented on packages/scripts/entrypoint_nestri.sh:
```diff
@@ -1,8 +1,12 @@
+else
+    log "Error: Common script not found at /etc/nestri/common.sh"
+    exit 1
+fi
```
Oops, good catch, should be resolved by next commit.

DatCaptainHorse commented on packages/scripts/entrypoint_nestri.sh:
```diff
@@ -156,8 +149,8 @@
+    # Start Steam patcher only if Steam command is present and running under non-sane container engines
+    if [[ -n "${NESTRI_LAUNCH_CMD}" ]] && [[ "$NESTRI_LAUNCH_CMD" == *"steam"* ]] && [[ "$container_runtime" != "podman" ]]; then
         start_steam_namespaceless_patcher
     fi
```
Good catch, resolving on next commit.

DatCaptainHorse commented on packages/server/Cargo.toml:
```diff
@@ -2,15 +2,16 @@
 name = "nestri-server"
 version = "0.1.0-alpha.2"
 edition = "2024"
+rust-version = "1.88"
```
Ah right, Rust did a change which uses that file now, I'll add in next commit, good catch.

DatCaptainHorse commented on packages/server/src/gpu.rs:
```diff
@@ -208,3 +271,21 @@
+            println!("{}", gpu);
+        }
+    }
+}
```
Yeah wasn't sure how to handle it, seems you have it figured out though, thanks. Should be resolved by next commit.

DatCaptainHorse commented on packages/scripts/entrypoint_nestri.sh:
```diff
@@ -259,33 +258,38 @@
+    # Make sure we have DBus
+    export $(dbus-launch)
+
     restart_chain
```
I suppose it doesn't hurt to do this, resolved by next commit.

DatCaptainHorse commented on packages/server/src/main.rs:
```diff
@@ -239,6 +228,10 @@ async fn main() -> Result<(), Box<dyn Error>> {
     // Handle audio encoder selection
     let audio_encoder = handle_encoder_audio(&args);
 
+    // Initialize libp2p (logically the sink should handle the connection to be independent)
+    let nestri_p2p = Arc::new(NestriP2P::new().await?);
+    let p2p_conn = nestri_p2p.connect(relay_url).await?;
```
in `../p2p/p2p.rs` the line `let conn_addr: Multiaddr = conn_url.parse()?;` should take care of this already no? It tries parsing multiaddr and if it's invalid or empty it returns an error. Resolve if I'm correct :pray:

DatCaptainHorse commented on packages/server/src/gpu.rs:
```diff
@@ -49,73 +63,122 @@ impl GPUInfo {
-        )
-        .collect()
+            });
+        }
```
Fixed for next commit, bit odd that I didn't get any errors or even warnings from this.

DatCaptainHorse commented on packages/server/src/main.rs:
```diff
@@ -313,6 +306,13 @@
+        "P010_10LE"
+    } else {
+        "NV12"
+    };
```
You don't happen to have recent knowledge on what AMD and NVIDIA name their P010 formats as in gstreamer, rabbit? I only got Intel GPUs myself so can't check myself unless I ask around.

DatCaptainHorse replied:
Hm, for now I'll keep it as is, since 8-bit is default and 10-bit is bit hit&miss with some browsers anyhow.
Remember/remind me later to add helpers for checking supported formats for an element perhaps. Resolved for now.

DatCaptainHorse commented on packages/scripts/container_helpers.sh:
```diff
@@ -0,0 +1,83 @@
+            fi
+            if grep -q "image=" "/run/.containerenv" 2>/dev/null; then
+                container_info["image"]=$(grep "image=" "/run/.containerenv" | cut -d'=' -f2- | tr -d '"')
+            fi
```
Done for next commit, fair enough catch :+1:

DatCaptainHorse committed:
Fix some nitpicks

- In addition to using new rust-toolchain.toml file, updated to Rust 1.89 as that's stable now.

DatCaptainHorse committed:
Fix nestri user pwd generation

- Rabbit was wrong, without me noticing immediately..

DatCaptainHorse commented:
@AquaWolf added you as reviewer, since you're more security sensitive than me :sweat_smile: - I added build-time password generation so there's no plain default "nestri1234". It's not exported anymore either so if you haven't built the image yourself (or haven't asked the one who built it), then the password is left unknown as it's not in ENV anymore.

AquaWolf commented:
> @AquaWolf added you as reviewer, since you're more security sensitive than me 😅 - I added build-time password generation so there's no plain default "nestri1234". It's not exported anymore either so if you haven't built the image yourself (or haven't asked the one who built it), then the password is left unknown as it's not in ENV anymore.

Hmm but in general that password is known in build time right? So if you enable ssh the password would be known by the one who built the image right?

DatCaptainHorse commented on containerfiles/runner.Containerfile:
```diff
@@ -121,58 +121,51 @@ RUN --mount=type=cache,target=${CARGO_HOME}/registry \
         pipewire pipewire-pulse pipewire-alsa wireplumber \
         noto-fonts-cjk supervisor jq chwd lshw pacman-contrib \
-        openssh && \
+        hwdata openssh \
```
It is key-only, resolved as you're touching things outside the scope, as hwdata package was added, openssh has already been there.

DatCaptainHorse commented:
> > @AquaWolf added you as reviewer, since you're more security sensitive than me 😅 - I added build-time password generation so there's no plain default "nestri1234". It's not exported anymore either so if you haven't built the image yourself (or haven't asked the one who built it), then the password is left unknown as it's not in ENV anymore.
> 
> Hmm but in general that password is known in build time right? So if you enable ssh the password would be known by the one who built the image right?

We only allow key-only SSH though, not password authenticated SSH, so it won't matter.

DatCaptainHorse commented on containerfiles/runner.Containerfile:
```diff
@@ -121,58 +121,51 @@ RUN --mount=type=cache,target=${CARGO_HOME}/registry \
+    echo "${NESTRI_USER}:${NESTRI_USER_PWD}" | chpasswd && \
+    mkdir -p ${NESTRI_XDG_RUNTIME_DIR} && \
+    chown ${NESTRI_USER} ${NESTRI_XDG_RUNTIME_DIR} && \
+    usermod -aG input,video,render,seat ${NESTRI_USER}
```
If we ever need that we'll do it, for now resolving because it's not used. SSH uses key-only auth.

DatCaptainHorse commented on containerfiles/runner.Containerfile:
```diff
@@ -86,7 +86,7 @@ RUN --mount=type=cache,target=/var/cache/pacman/pkg \
 
 # Clone repository
-RUN git clone --depth 1 -b "dev-dmabuf" https://github.com/DatCaptainHorse/gst-wayland-display.git
+RUN git clone --depth 1 -b "stable" https://github.com/games-on-whales/gst-wayland-display.git
```
Doing in next commit, also as of note `--rev` parameter exists, there's no need to do cd'ing and all that..

DatCaptainHorse commented:
Marking as draft since I want to make more changes without rabbit activating for time being.

DatCaptainHorse committed:
Nitpick fixes, cleaning up and solidifying changes.

DatCaptainHorse committed:
Add missing loopback module config

DatCaptainHorse commented on packages/scripts/container_helpers.sh:
```diff
@@ -0,0 +1,77 @@
+            echo "unknown"
+        fi
+    fi
+}
```
Good catch, fixing by next commit..

DatCaptainHorse commented on packages/scripts/entrypoint.sh:
```diff
@@ -1,20 +1,24 @@
-    local user_group="${USER}:${GID}"
-    if ! chown -h --no-preserve-root "$user_group" "${HOME}" 2>/dev/null; then
-        echo "Error: Failed to change ownership of ${HOME} to ${user_group}" >&2
+    if ! $ENTCMD_PREFIX chown "${NESTRI_USER}:${NESTRI_USER}" "${NESTRI_HOME}" 2>/dev/null; then
+        echo "Error: Failed to change ownership of ${NESTRI_HOME} to ${user_group}" >&2
         return 1
```
Oops, fixing by next commit.

DatCaptainHorse commented on packages/server/src/args/device_args.rs:
```diff
@@ -1,37 +1,52 @@
-                .get_one::<String>("gpu-name")
-                .unwrap_or(&"".to_string())
-                .clone(),
-            gpu_index: matches.get_one::<i32>("gpu-index").unwrap_or(&-1).clone(),
+                .get_one::<Option<String>>("gpu-name")
+                .cloned()
+                .unwrap_or(None),
+            gpu_index: matches
+                .get_one::<Option<u32>>("gpu-index")
+                .cloned()
+                .unwrap_or(None),
             gpu_card_path: matches
-                .get_one::<String>("gpu-card-path")
-                .unwrap_or(&"".to_string())
-                .clone(),
+                .get_one::<Option<String>>("gpu-card-path")
+                .cloned()
+                .unwrap_or(None),
         }
```
Gah, never allow me to code when tired, it literally shows in IDE as "Option<Option<String>>" which is bad. Fixing by next commit..

DatCaptainHorse commented on packages/server/src/enc_helper.rs:
```diff
@@ -375,22 +375,25 @@ pub fn get_compatible_encoders(gpus: &Vec<GPUInfo>) -> Vec<VideoEncoderInfo> {
                         }
-                        EncoderAPI::NVENC if element.has_property("cuda-device-id", None) => {
+                        EncoderAPI::NVENC if element.has_property("cuda-device-id") => {
                             let cuda_id = element.property::<u32>("cuda-device-id");
                             get_nvidia_gpu_by_cuda_id(&gpus, cuda_id as usize)
                         }
```
You're bit unclear with your wording there, what the heck is DeepStream? You're going off the rails. Anyhow, I get what you're saying, and tbh I hate relying on `nvidia-smi` for the method that checks anyhow so I'll go by the `nvh264deviceNenc` way.

DatCaptainHorse committed:
Nitpicks and fixes

- Also add missing supervisord.conf changes..

DatCaptainHorse commented on packages/scripts/entrypoint.sh:
```diff
@@ -209,6 +217,20 @@ configure_ssh() {
 
 main() {
+    # Wait for required sockets
+    wait_for_socket "${NESTRI_XDG_RUNTIME_DIR}/dbus-1" "DBus" || exit 1
+    wait_for_socket "${NESTRI_XDG_RUNTIME_DIR}/pipewire-0" "PipeWire" || exit 1
+
```
In some setups? We are sticking with CachyOS/Arch for our runner container so I doubt we'll see "some setups". Also, it is in-fact `dbus-1` in our containerized setup, it is either a symlink to `system_bus_socket` or the actual socket (if running it under user rather than root).

Resolving.

DatCaptainHorse commented on packages/scripts/entrypoint.sh:
```diff
@@ -209,6 +217,20 @@ configure_ssh() {
+    }
+    log_container_info
+
+    if [[ "$container_runtime" != "apptainer" ]]; then
+        ENTCMD_PREFIX="sudo -E"
+    fi
+
```
> if sudo isn’t present

sudo is ALWAYS present, we are running in container!!! Try to keep up rabbit. Resolving.

DatCaptainHorse commented on packages/server/src/args/encoding_args.rs:
```diff
@@ -105,9 +106,9 @@ impl VideoEncodingOptions {
-                    .get_one::<String>("video-encoder")
-                    .unwrap_or(&"".to_string())
-                    .clone(),
+                    .get_one::<Option<String>>("video-encoder")
+                    .cloned()
+                    .unwrap_or(None),
                 rate_control: match matches
```
Missed that one, fixing in next commit.

DatCaptainHorse commented on packages/server/src/main.rs:
```diff
@@ -25,38 +25,32 @@ use tracing_subscriber::filter::LevelFilter;
+        if let Some(gpu) = gpu::get_gpu_by_card_path(&gpus, gpu_card_path.as_str()) {
             return Ok(Vec::from([gpu]));
         }
     } else {
```
Fair catch, resolving in next commit.

DatCaptainHorse committed:
Hopefully last nitpicks

DatCaptainHorse committed:
Last nitpicks

DatCaptainHorse commented on packages/server/src/main.rs:
```diff
@@ -250,7 +249,9 @@ async fn main() -> Result<(), Box<dyn Error>> {
         }
         encoding_args::AudioCaptureMethod::PIPEWIRE => {
-            gstreamer::ElementFactory::make("pipewiresrc").build()?
+            let pw_element = gstreamer::ElementFactory::make("pipewiresrc").build()?;
+            pw_element.set_property("use-bufferpool", &false); // false for audio
+            pw_element
         }
```
Oh my god rabbit, we are not doing multi-distro containers, just single one, CachyOS which is based on Arch. Get a grip, resolving as it's not an issue..

DatCaptainHorse commented on packages/scripts/envs.sh:
```diff
@@ -1,9 +1,10 @@
 #!/bin/bash
 
-export XDG_RUNTIME_DIR=/run/user/${UID}/
+export LANG=${NESTRI_LANG}
+export HOME=${NESTRI_HOME}
+export XDG_RUNTIME_DIR=${NESTRI_XDG_RUNTIME_DIR}
```
Set in container.. resolving.

DatCaptainHorse commented on packages/scripts/supervisord.conf:
```diff
@@ -1,24 +1,15 @@
 priority=1
-
-[program:seatd]
-user=root
-command=seatd
-autorestart=true
-autostart=true
-startretries=3
-priority=2
+environment=XDG_RUNTIME_DIR=%(ENV_NESTRI_XDG_RUNTIME_DIR)s
```
This is a conscious decision, as apptainer won't work in root mode, it will inherit the user and DBUS will work only that way, if it's started by root user, it will function outside apptainer still. Resolving.

DatCaptainHorse commented on packages/scripts/supervisord.conf:
```diff
@@ -1,24 +1,15 @@
-autostart=true
-startretries=3
-priority=2
+environment=XDG_RUNTIME_DIR=%(ENV_NESTRI_XDG_RUNTIME_DIR)s
```
Sigh, it exists in containerfile.
